### PR TITLE
Flux and ReSwift 

### PR DESCRIPTION
**Pull request by devladinci**
## Scope 😻🚀
- [x] Use ReSwift
- [x] Migrate to new App State
- [x] Remove `Notifier`, `PHUserDefaults` and etc.
## Notes
- More about [ReSwift](https://github.com/ReSwift/ReSwift)
- More about [Redux](https://github.com/reactjs/redux)
- More about [Flux](https://facebook.github.io/react/docs/flux-overview.html)

### CODE REVIEW

File: Source/PHTokenOperation.swift [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+//  Product Hunt
+//
+//  Created by Vlado on 5/3/16.
+//  Copyright © 2016 ProductHunt. All rights reserved.
+//
+
+import Foundation
+import ReSwift
+
+// TODO: Remove if not needed
```
Comment (RStankov) [inline Source/PHTokenOperation.swift]: Is it needed ?

File: Source/PHTokenOperation.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+class PHTokenOperation {
+
+    class func perform() {
+        perform(store, api: PHAPI.sharedInstance)
+    }
+
+    class func perform(store: Store<PHAppState>, api: PHAPI) {
```
Comment (RStankov) [inline Source/PHTokenOperation.swift]: It is better to have this be `perform<T>(store: Store<T>, api: PHAPI)`.

File: Source/PHTokenModule.swift [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+//  Product Hunt
+//
+//  Created by Vlado on 5/3/16.
+//  Copyright © 2016 ProductHunt. All rights reserved.
+//
+
+import Foundation
+import ReSwift
+
+struct PHTokenGet: Action {
```
Comment (RStankov) [inline Source/PHTokenModule.swift]: Maybe a better name will be `PHTokenSetAction`, to be consistent with other actions.

File: Source/Models/PHDefaults.swift [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    private let kSeenPostsKey = "seenPostsKey"
+    private let kSeenPostsKeyDate = "seenPostsDate"
+    private let kSeenPostsKeyIds = "seenPostsDateIds"
+
+    private let kTokenKey = "oauthToken"
+    private let kTokenKeyDate = "expirationDate"
+    private let kTokenKeyAccess = "accessToken"
+
+    init() {
```
Comment (RStankov) [inline Source/Models/PHDefaults.swift]: I think store should be passed as argument where possible.

File: Source/Controllers/PHPostListViewController.swift [hunk 5]
```diff
@@ -10,55 +10,55 @@ import Cocoa
...
     override func viewWillAppear() {
         super.viewWillAppear()
 
-        updateUITimer = NSTimer.every(15.seconds) { [weak self] in self?.updateUI() }
-
         self.source.loadNewer()
 
         PHScheduleAsSeenAction.performCancel()
+
+        updateUITimer = NSTimer.every(15.seconds) { [weak self] in self?.updateUI() }
```
Comment (RStankov) [inline Source/Controllers/PHPostListViewController.swift:47]: Is this still needed? Since now all state changes are going through the store.

File: Source/Models/API/PHAPIEndpoint.swift [hunk 6]
```diff
@@ -27,7 +27,7 @@ class PHAPIEndpoint {
         updateAuthorizationToken()
     }
 
-    func updateAuthorizationToken(token: PHToken? = PHKeychain.getToken()) {
+    func updateAuthorizationToken(token: PHToken? = store.state.token) {
```
Comment (RStankov) [inline Source/Models/API/PHAPIEndpoint.swift]: Can we always pass the token. Making `PHAPIEndpoint` depend on a global variable is problematic.

File: Source/Models/PHSettings.swift [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+//  PHSettings.swift
+//  Product Hunt
+//
+//  Created by Vlado on 4/27/16.
+//  Copyright © 2016 ProductHunt. All rights reserved.
+//
+
+import Foundation
+
+struct PHSettings {
```
Comment (RStankov) [inline Source/Models/PHSettings.swift]: I think store state structs should be in their respected module file.

File: Source/Models/ViewModels/PHPostViewModel.swift [hunk 8]
```diff
@@ -6,12 +6,16 @@
...
 
 class PHPostViewModel {
 
     var isSeen: Bool {
-        return PHSeenPosts.isSeen(post)
+        if PHDateFormatter.daysAgo(post.day) > 0 {
+            return true
+        }
+
+        return  store.state.seenPosts.postIds.contains(post.id)
```
Comment (RStankov) [inline Source/Models/ViewModels/PHPostViewModel.swift]: `isSeen` from `PHSeenPosts` ?

Comment (devladinci) [inline Source/Models/ViewModels/PHPostViewModel.swift]: 👍

File: Source/Controllers/PHPostListViewController.swift (hunk 5 above)
Comment (devladinci) [inline Source/Controllers/PHPostListViewController.swift:47]: Yes, this updates `timeAgo` label only. Changes in the store does not occur so often and label stops updating.

File: Source/Models/PHDefaults.swift (hunk 4 above)
Comment (devladinci) [inline Source/Models/PHDefaults.swift]: 👍

File: Source/PHTokenOperation.swift (hunk 1 above)
Comment (devladinci) [inline Source/PHTokenOperation.swift]: Good reminder, actually it wasn't

File: Source/Models/PHSettings.swift (hunk 7 above)
Comment (devladinci) [inline Source/Models/PHSettings.swift]: For me having `Actions`, `Reducer` and `state structure` in one `module` file can be hard to follow

File: Source/PHTokenModule.swift (hunk 3 above)
Comment (devladinci) [inline Source/PHTokenModule.swift]: 👍

**Comment by devladinci** [comment]
Merged.
